GitHub workflow for running unit tests for MKL backend on CPU

rscohn2 opened the pull request:
# Description

GitHub Actions workflow to run unit tests for MKL/CPU. 

I disabled tests that require GPU by excluding tests where the name ends in `gpu`. RNG and DFT domains finish in 10 minutes. I am still waiting for BLAS/LAPACK to finish. We could have nightly testing that runs everything and PR testing that runs a smaller set. Any thoughts on how to reduce the testing for PR's?

@mkrainiuk

rscohn2 committed:
pr testing

rscohn2 commented:
lapack finishes in 30 minutes and blas takes 1.5 hours. All the time is building the tests. Combining multiple tests into a single binary might make testing faster without reducing coverage.

rscohn2 committed:
add cpu to job name

rscohn2 commented:
@hjabird: Added CPU to job name to address: https://github.com/oneapi-src/oneMKL/pull/482#discussion_r1598245390

rscohn2 committed:
use intel runners to avoid 20 runner limit for github provided runners

sknepper commented on .github/workflows/pr.yml:
```diff
@@ -0,0 +1,58 @@
+env:
+  CTEST_OUTPUT_ON_FAILURE: 1
+  LAPACK_VERSION: 3.12.0
+  PARALLEL: -j 2
```
What sort of system will this run on? If we can increase the number of parallel jobs, that should significantly decrease the build time of tests.

rscohn2 replied:
github provided runners have 1 core, 2 threads. You can have up to 20 concurrent jobs. There is a way to pay for runners with more cores. I don't think it would be expensive. Someone is looking into getting a dedicated node in devcloud that we could use. Then 1 runner could use 120 cores and 4 GPU's. This would solve most of our problems so I am hoping we could live with slow performance until that happens.

rscohn2 replied:
> What sort of system will this run on? If we can increase the number of parallel jobs, that should significantly decrease the build time of tests.

The new github org: `onemkl-project` has runners for a 24 core system. It reduces the total time to 30 minutes. Still working on PVC and more cores.
https://github.com/onemkl-project/ci-test/actions/runs/9140509510

sknepper commented on .github/workflows/pr.yml:
```diff
@@ -0,0 +1,58 @@
+        cmake ${SHARED_OPT} -DBUILD_INDEX64=on -B lapack/build64
+        cmake --build lapack/build64 ${PARALLEL} --target install
+    - name: Install oneapi
+      uses: rscohn2/setup-oneapi@2ad0cf6b74bc2426bdcee825cf88f9db719dd727 # v0.1.0
```
It looks like we'll need to update this workflow for every public release of mkl/icx. Will this uses line also need to be updated each time the version of mkl/icx changes? Is there a way to somehow signify "latest" that doesn't require regularly updating this workflow? Thanks!

rscohn2 replied:
`rscohn2/setup-oneapi@v0`
would pick up the latest. I use hashes because I believe the openssf scanning tool flags anything that is not a hash.  After we move the repo to onemkl-project org, we could decide to allow versions instead of hashes. Pinning the CI to a specific compiler/MKL makes CI more stable/reproducible so I recommend leaving it as is.

mkrainiuk replied:
Should we use faceless account for such automation?

rscohn2 replied:
My id appears there because that is the URL for the plugin that is used to install oneapi components:
https://github.com/rscohn2/setup-oneapi
The plugin simplifies the install and reduces the install time from 5 minutes to 30 seconds.

mkrainiuk replied:
I understand why we need it, but as far as I know using personal id could be a problem for security scans, but I'm not sure if it's a case here.

rscohn2 replied:
I took a couple repos through the openssf badge process which includes openssf best practices. There were several requirements related to github actions, but I don't think there is anything about which org a GitHub Action comes from. The rules for private repos with access to intel resources are different, but this is a public repo. In any case, if there is a problem in the future, I will fix it.

dnhsieh-intel commented on .github/workflows/pr.yml:
```diff
@@ -0,0 +1,58 @@
+
+jobs:
+  mkl:
+    runs-on: intel-ubuntu-latest
```
We run CPU tests on Red Hat 9 internally. If there are no issues, we could use this as an extended testing.

rscohn2 replied:
This is public testing for PR's. Non-intel contributors and reviewers will be able to see and correct problems before they are merged. It is limited by compute resources available: cpu only, single core runners, etc. The intel testing is needed to provide extensive coverage. but requires intel people to look at the results.

dnhsieh-intel replied:
I understand the scope and purpose of this public CI. I was thinking it might be a little confusing that we list only Red Hat 9 in [the supported operating systems](https://github.com/oneapi-src/oneMKL?tab=readme-ov-file#supported-operating-systems) for CPU but run tests on Ubuntu.

rscohn2 replied:
Thanks for clarifying.

rscohn2 committed:
ubuntu-latest runners are faster, but there are more intel-ubuntu-latest runner available

mkrainiuk commented:
> Any thoughts on how to reduce the testing for PR's? 

I guess we can also skip building GPU part with `-DENABLE_MKLGPU_BACKEND=False` for now since it's not tested anyway, so it will save some build time, and we can run only runtime dispatching tests and ignore compiler time dispatching tests with `-E /CT/`, it should give us significant testing time reduction.

rscohn2 committed:
github does not support runs-on targetting multiple labels

rscohn2 committed:
disable gpu and ct tests in pr testing

rscohn2 commented:
> > Any thoughts on how to reduce the testing for PR's?
> 
> I guess we can also skip building GPU part with `-DENABLE_MKLGPU_BACKEND=False` for now since it's not tested anyway, so it will save some build time, and we can run only runtime dispatching tests and ignore compiler time dispatching tests with `-E /CT/`, it should give us significant testing time reduction.

I made the change. It runs about 25% faster, which makes up for the intel hosted runners being 25% slower than github hosted runners. When you move to onemkl-project org, we can use github hosted runners without contending with other projects.

dnhsieh-intel commented on .github/workflows/pr.yml:
```diff
@@ -0,0 +1,58 @@
+    - name: Configure/Build for a domain
+      run: |
+        source /opt/intel/oneapi/setvars.sh
+        cmake -DREF_BLAS_ROOT=${PWD}/lapack/install -DREF_LAPACK_ROOT=${PWD}/lapack/install -DENABLE_MKLGPU_BACKEND=off -DTARGET_DOMAINS=${{ matrix.domain }} -B build
```
A verbose build output could help debugging.
```suggestion
        cmake -DREF_BLAS_ROOT=${PWD}/lapack/install -DREF_LAPACK_ROOT=${PWD}/lapack/install -DENABLE_MKLGPU_BACKEND=off -DTARGET_DOMAINS=${{ matrix.domain }} -DCMAKE_VERBOSE_MAKEFILE=on -B build
```

dnhsieh-intel reviewed (approved):
Thanks for the PR! I have one suggestion.

dnhsieh-intel commented on .github/workflows/pr.yml:
```diff
@@ -0,0 +1,58 @@
+      run: |
+        source /opt/intel/oneapi/setvars.sh
+        # Run tests, skip GPU tests
+        ctest --test-dir build ${PARALLEL} -E 'ct$'
```
And I would suggest using `-j 1` for testing:
```suggestion
        ctest --test-dir build -j 1 -E 'ct$'
```
If we run in parallel, I am worried that the log would be hard to read when there are failures.

rscohn2 committed:
Apply suggestions from code review

verbose logging and serial test execution for more readable logs. build time dominates so parallel test is less important.

Co-authored-by: Dai-Ni Hsieh <dai-ni.hsieh@intel.com>
